reactor-netty-http-brave/src/main/java/reactor/netty/http/brave/TracingHttpServerDecorator.java

```diff
@@ -0,0 +1,277 @@
+import brave.propagation.CurrentTraceContext;
+import brave.propagation.TraceContext;
+import io.netty.channel.EventLoop;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.SignalType;
+import reactor.netty.Connection;
+import reactor.netty.ConnectionObserver;
+import reactor.netty.http.brave.internal.InternalPropagation;
+import reactor.netty.http.server.DelegatingHttpServerOperations;
+import reactor.netty.http.server.HttpServer;
+import reactor.util.annotation.Nullable;
+
+import java.net.InetSocketAddress;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+
+import static java.util.Objects.requireNonNull;
+import static reactor.netty.ConnectionObserver.State.CONFIGURED;
+import static reactor.netty.http.server.HttpServerState.REQUEST_DECODING_FAILED;
+
+final class TracingHttpServerDecorator {
+
+	static final DelegatingHttpServerOperations instance = DelegatingHttpServerOperations.INSTANCE;
+
+	final CurrentTraceContext currentTraceContext;
+	final HttpServerHandler<HttpServerRequest, HttpServerResponse> handler;
+	final Function<String, String> uriMapping;
+
+	TracingHttpServerDecorator(HttpTracing httpTracing, Function<String, String> uriMapping) {
+		requireNonNull(httpTracing, "httpTracing");
+		this.currentTraceContext = httpTracing.tracing().currentTraceContext();
+		this.handler = HttpServerHandler.create(httpTracing);
+		this.uriMapping = requireNonNull(uriMapping, "uriMapping");
+	}
+
+	HttpServer decorate(HttpServer server) {
+		return server.childObserve(new TracingConnectionObserver(currentTraceContext, handler, uriMapping));
+	}
+
+	static final class DelegatingHttpRequest extends HttpServerRequest {
+
+		final reactor.netty.http.server.HttpServerRequest delegate;
+		final Function<String, String> uriMapping;
+
+		DelegatingHttpRequest(reactor.netty.http.server.HttpServerRequest delegate, Function<String, String> uriMapping) {
+			this.delegate = delegate;
+			this.uriMapping = uriMapping;
+		}
+
+		@Override
+		@Nullable
+		public String header(String name) {
+			requireNonNull(name, "name");
+			return delegate.requestHeaders().get(name);
+		}
+
+		@Override
+		public String method() {
+			return delegate.method().name();
+		}
+
+		@Override
+		public boolean parseClientIpAndPort(Span span) {
+			requireNonNull(span, "span");
+			//HttpServerOperations handles Forwarded/X-Forwarded and proxy protocol (HAProxy)
+			InetSocketAddress remoteAddress = delegate.remoteAddress();
+			if (remoteAddress == null) {
+				// This can happen only in case of UDS or bad request
+				return false;
+			}
+			return span.remoteIpAndPort(remoteAddress.getHostString(), remoteAddress.getPort());
+		}
+
+		@Override
+		@Nullable
+		public String path() {
+			try {
+				return delegate.fullPath();
+			}
+			catch (IllegalStateException e) {
+				return null;
+			}
+		}
+
+		@Override
+		@Nullable
+		public String route() {
+			return path() == null ? null : uriMapping.apply(path());
```

Comment:
how costly is the `path()` call? consider extracting in a variable to call the method once